src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -2188,6 +2191,7 @@ static void on_complete(void* arg, grpc_error* error) {
                                                   &closures);
     }
   }
+  calld->call_combiner->TsanUnlock();
```

Comment:
How do we know that this is the right place to put this?  At first glance, I would think it needs to go further down, because we are still accessing data in `calld` that is protected by the call combiner.  (Technically, we should probably do this inside of `CallCombinerClosureList::RunClosures()`.)

Note that I am asking here less in terms of how we know this is right in this specific case and more in terms of how an engineer seeing this in the future will decide where to put it.  Or, to say this another way, we need to improve the documentation on the `TsanLock()` and `TsanUnlock()` API so that future engineers can figure out how it's actually supposed to be used.

Reply:
The problem with `CallCombinerClosureList::RunClosures()` is that we schedule the closure and do not directly run it. With that, unfortunately, annotating around `GRPC_CLOSURE_SCHED()` wouldn't fix the tsan error.

Reading your feedback, it occurred to me that this is really not going to scale and it's going to be very confusing for engineers seeing this in the future. I agree we should take care of this transparently, which I think is done in the new version on #17368